nexus/src/app/background/vpc_routes.rs

```diff
@@ -0,0 +1,279 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for propagating VPC routes (system and custom) to sleds.
+
+use super::common::BackgroundTask;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_model::{Sled, SledState, Vni};
+use nexus_db_queries::{context::OpContext, db::DataStore};
+use nexus_networking::sled_client_from_address;
+use nexus_types::{
+    deployment::SledFilter, external_api::views::SledPolicy, identity::Asset,
+    identity::Resource,
+};
+use omicron_common::api::internal::shared::{
+    ResolvedVpcRoute, ResolvedVpcRouteSet, RouterId, RouterVersion,
+};
+use serde_json::json;
+use std::collections::hash_map::Entry;
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+};
+use uuid::Uuid;
+
+pub struct VpcRouteManager {
+    datastore: Arc<DataStore>,
+}
+
+impl VpcRouteManager {
+    pub fn new(datastore: Arc<DataStore>) -> Self {
+        Self { datastore }
+    }
+}
+
+// This RPW doesn't concern itself overly much with resolved router targets
+// and destinations being partial wrt. the current generation, in the same
+// vein as how firewall rules are handled. Gating *pushing* this update on a
+// generation number can be a bit more risky, but there's a sort of eventual
+// consistency happening here that keeps this safe.
+//
+// Any location which updates name-resolvable state follows the pattern:
+//  * Update state.
+//  * Update (VPC-wide) router generation numbers.
+//  * Awaken this task. This might happen indirectly via e.g. instance start.
```

Comment:
Dumb Q: This task is only woken/awaken on calls to [`vpc_router_increment_rpw_version`](https://github.com/oxidecomputer/omicron/pull/5777/files#diff-0017315ba1221fe0d101b6602d359187f4b40e5c3a79b8aa2be94bd85758b3fcR292) right? Or directly via events like "on" vpc subnet creation through `vpc_needed_notify_sleds`?

Reply:
It's not super obvious, but all background tasks are also woken up periodically by all Nexus instances, at a period they choose when they're initialized: https://github.com/oxidecomputer/omicron/pull/5777/files#diff-b9fd8b96da4cd0201ab8cb1f87086d4fba616d9cb08276ab766e6c63a3ee0932R416

Reply:
> woken up periodically by all Nexus instances

Sorry, let me clarify this: each Nexus instance is running the same set of background tasks, and each instance _independently_ wakes its own tasks up periodically. (In practice it's common for the tasks on all three Nexus instances to wake up close to the same time because the Nexus instances are often started at about the same time, but that isn't guaranteed / drifts over time / etc. of course.)

Reply:
This particular task doesn't define any events that trigger it, but the background task subsystem does allow you to [register watchers](https://github.com/oxidecomputer/omicron/blob/d2af4e414d989f088e6078d81220fcfe6ec9e43e/nexus/src/app/background/common.rs#L230), so that tasks can communicate with each other if needed.

Reply:
@bnaecker gotcha, awesome. Makes sense. So, in https://github.com/oxidecomputer/omicron/pull/5777/files#diff-0017315ba1221fe0d101b6602d359187f4b40e5c3a79b8aa2be94bd85758b3fcR292, does the version increment not awaken this RPW, as per the comment? Or does it refer to another one?

Reply:
It does look like this calls `vpc_router_increment_rpw_version()`, which then calls `vpc_needed_notify_sleds()`, which explicitly activates the task.

https://github.com/oxidecomputer/omicron/blob/e886d167b33a8991a2a17be43a4ed12fe586fd57/nexus/src/app/vpc_router.rs#L280-L286

There are really three ways for a task to activate:

- A timer firing (this is the most common path, in my experience).
- An "upstream" task notified this task to activate, so that we can define trees of tasks.
- Someone can explicitly activate it, like we have here.

Reply:
@bnaecker Yep, I followed the explicit activation (but missed the call to `vpc_router_increment_rpw_version` within the background task here I guess?) Anyhow, good to know, particularly on the second option as well with the tree of tasks. 

Thanks!